{"cells":[{"metadata":{},"cell_type":"markdown","source":"# COMP 364: Sequence Handling with BioPython\n\n\n## Warmup: OOP Practice\n\nWrite a class that representas a Student with two attributes: name, and grade. Next, write a class called Course that represents the students in the course with two initial attribute: course name, and a list of Students, a max capacity (class attribute). The Course class can also compute the average of all the students grades and add a student to the course."},{"metadata":{"trusted":true},"cell_type":"code","source":"!pip install Biopython","execution_count":7,"outputs":[{"output_type":"stream","text":"WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.\nPlease see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.\nTo avoid this problem you can invoke Python with '-m pip' instead of running pip directly.\nRequirement already satisfied: Biopython in /srv/conda/envs/notebook/lib/python3.7/site-packages (1.78)\nRequirement already satisfied: numpy in /srv/conda/envs/notebook/lib/python3.7/site-packages (from Biopython) (1.20.1)\n","name":"stdout"}]},{"metadata":{"trusted":true},"cell_type":"code","source":"import statistics\n\nclass Student:\n    def __init__(self, name, grade):\n        self.name = name\n        self.grade = grade\nclass Course:\n    max_capacity = 3\n    def __init__(self, name):\n        self.course_name = name\n        self.students = []\n    def add_student(self, student_name, grade):\n        if len(self.students) + 1 > Course.max_capacity:\n            print(\"Class is full\")\n        else:\n            self.students.append(Student(student_name, grade))\n    def class_avg(self):\n        return statistics.mean([s.grade for s in self.students])\n\n\n\nc = Course(\"COMP 364\")\nprint(c.course_name)\nprint(c.students)\n\nc.add_student(\"Carlos\", 3.4)\nc.add_student(\"Chris\", 4.0)\nc.class_avg()\n\nb = Course(\"COMP 202\")\nb.add_student(\"Carlos\", 1.2)\n\nprint([student.name for student in b.students])\nprint([student.name for student in c.students])","execution_count":8,"outputs":[{"output_type":"stream","text":"COMP 364\n[]\n['Carlos']\n['Carlos', 'Chris']\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"# BioPython\n\nBioPython is a very popular third party Python package for handling biological data.\n\nIn order to install:\n\n```\nconda install biopython\n```\n\nBioPython has three major functionalities:\n\n* Sequence Handling\n* 3D Structure\n* Population Genetics\n\nToday we'll cover sequence handling and next time we will do 3D structure handling, and if there is time we may do some Population Genetics.\n\nSome useful references: [tutorial](http://biopython.org/DIST/docs/tutorial/Tutorial.html), [website](http://biopython.org/), [wiki](http://biopython.org/wiki/Category%3AWiki_Documentation) (most of my examples are coming from these sources).\n \n# `BioPython.Seq`\n\nThe main object we'll be dealing with are sequences. This is handled by the `Seq` class.\n\nAs we know, we can have DNA, RNA, and Protein sequences.\n\nBioPython helps us cleanly distinguish and do different things with different kinds of sequences."},{"metadata":{"trusted":true},"cell_type":"code","source":"#let's make a generic sequence\n\nfrom Bio.Seq import Seq\n\nmy_seq = Seq(\"CCCGGAGAGA\")\nprint(type(my_seq))\n#let's see what attributes this object has\nattributes = [a for a in dir(my_seq) if not a.startswith(\"_\")]\nprint(attributes)","execution_count":9,"outputs":[{"output_type":"stream","text":"<class 'Bio.Seq.Seq'>\n['back_transcribe', 'complement', 'complement_rna', 'count', 'count_overlap', 'encode', 'endswith', 'find', 'index', 'join', 'lower', 'lstrip', 'reverse_complement', 'reverse_complement_rna', 'rfind', 'rindex', 'rsplit', 'rstrip', 'split', 'startswith', 'strip', 'tomutable', 'transcribe', 'translate', 'ungap', 'upper']\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"It seems like there is a lot we can do with this object.\n\nThe problem is at this point Python doesn't know what kind of sequence this is (DNA, RNA, Protein).\n\nWe have to specify what kind of \"Alphabet\" the sequence belongs to."},{"metadata":{"trusted":true},"cell_type":"code","source":"#right now it has just a generic alphabet\nprint(my_seq.alphabet)","execution_count":10,"outputs":[{"output_type":"error","ename":"AttributeError","evalue":"'Seq' object has no attribute 'alphabet'","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)","\u001b[0;32m<ipython-input-10-7699f277e87d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#right now it has just a generic alphabet\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmy_seq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malphabet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mAttributeError\u001b[0m: 'Seq' object has no attribute 'alphabet'"]}]},{"metadata":{"trusted":true},"cell_type":"code","source":"from Bio.Alphabet import generic_dna, generic_protein, generic_rna\n\nmy_dna = Seq(\"CCCGGAGAG\", generic_dna)\nmy_rna = Seq(\"ACCCGUUGU\", generic_rna)\nmy_protein = Seq(\"AKKKGGGUUULL\", generic_protein)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Biopython will now know the difference for example between a DNA base `A` for adenine and the protein residue `A` for alanine.\n\nNow we can do some cool things.\n\nWe can perform the main actions of the central dogma: transcribe and translate."},{"metadata":{"trusted":true},"cell_type":"code","source":"my_gene = Seq(\"ACTAGCAGCGGA\", generic_dna)\n\n#get the mRNA\n\nmy_transcript = my_gene.transcribe()\nprint(my_transcript)\nprint(my_transcript.alphabet)\n\n\n#get the protein from the mRNA\nmy_protein = my_transcript.translate()\nprint(my_protein)\nprint(my_protein.alphabet)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"You can also go straight from the DNA to the protein."},{"metadata":{"trusted":true},"cell_type":"code","source":"coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\", generic_dna)\nmyprot = coding_dna.translate()\nprint(myprot)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"As you can see, we got some STOP codons represented as `*` and translation continued.\n\nWe can get translation to actually stop when it encounters a STOP codon."},{"metadata":{"trusted":true},"cell_type":"code","source":"myprot = coding_dna.translate(to_stop=True)\nprint(myprot)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We can be even more realistic and only allow translation of valid genes (i.e. with a valid start and stop codon and proper number of bases).\n\nThis is done by setting the `cds=True` keyword argument, which stands for \"coding sequence\".\n\nIf we don't have a valid coding sequence, we get an exception."},{"metadata":{"trusted":true},"cell_type":"code","source":"myprot = coding_dna.translate(cds=True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"gene = Seq(\"ATGGCCATTGTAATGTAG\", generic_dna)\ngene.translate(cds=True)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### General sequence methods\n\nThere are some convenient operations for dealing with sequences."},{"metadata":{},"cell_type":"markdown","source":"We can concatenate sequences if they are of matching type."},{"metadata":{"trusted":true},"cell_type":"code","source":"seq1 = Seq(\"AAACGGA\", generic_dna)\nseq2 = Seq(\"GGAGAT\", generic_dna)\n\nseq1 + seq2","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We can also index and slice as though we had strings."},{"metadata":{"trusted":true},"cell_type":"code","source":"seq1[:2] + seq2[-1]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"`Seq` objects are immutable, just like strings."},{"metadata":{"trusted":true},"cell_type":"code","source":"seq1[2] = \"G\"","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"There is another type of object called a `MutableSeq`. If we want to support mutability we can convert a `Seq` object to a `MutableSeq` object quite easily."},{"metadata":{"trusted":true},"cell_type":"code","source":"mut_seq = seq1.tomutable()\nmut_seq","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"mut_seq[0] = \"G\"\nprint(mut_seq)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We can also do searching inside sequences."},{"metadata":{"trusted":true},"cell_type":"code","source":"myseq = Seq(\"CCAGAAACCCGGAA\", generic_dna)\n\n#find the first occurence of the pattern\nprint(myseq.find(\"GAA\"))\n\n#find the number of non-overlapping occurrences of a pattern\nprint(myseq.count(\"GAA\"))","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Database Searching\n\nSay you isolated some piece of DNA in the lab and obtained its sequence.\n\nYou now want to find out which organism that sequence belongs to.\n\nYou ask a biologist and they tell you to just \"BLAST it\".\n\nAs you may know already, this is essentially the same as \"Google it\" for biologists.\n\nBLAST is an alignment algorithm that searches for your sequence of interest in a huge database of sequences whose origins are known.\n\nIf you didn't know BioPython, you would take your sequence `AAAAGGAGAGAGAGTTTATA` and go to the [NCBI BLAST web server](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome) and click on buttons like a monkey.\n\nThanks to BioPython we can do this programatically!\n\nThe `qblast` method from the `Bio.Blast.NCBIWWW` module essentially sends our sequence to the BLAST web server.\n\nHere we are using the \"nucleotice BLAST\" algorithm so we say `blastn` and we are using it on the database of all nucleotide sequences, called `nt`.\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"from Bio.Blast import NCBIWWW\n\nresult_handle = NCBIWWW.qblast(\"blastn\", \"nt\", Seq(\"AAAAGGAGAGAGAGTTTATA\", generic_dna))","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We wait a few seconds and get a `result_handle` which is like a temporary open file that we can read from.\n\nThe format of this file is in XML so not easy to read, thankfully BioPython has an XML parser that extracts all the information for us."},{"metadata":{"trusted":true},"cell_type":"code","source":"from Bio.Blast import NCBIXML\nblast_records = NCBIXML.parse(result_handle)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We get an \"iterator\" of BLAST record objects, or \"search results\". We can now loop over each of our search results and print some information.\n\nHere I am looping over all the results which each have an attribute `alignments` which are the alignments of our query sequence to some organism in the database.\n\nThe `alignment` attribute itself has other attributes like the `query` sequence, the `length` and `title` of the matching organism.."},{"metadata":{"trusted":true},"cell_type":"code","source":"for b in blast_records:\n    for alignment in b.alignments:\n        for hsp in alignment.hsps:\n            print('****Alignment****')\n            print('sequence:', alignment.title)\n            print('length:', alignment.length)\n            print('e value:', hsp.expect)\n            print(hsp.query[0:75] + '...')\n            print(hsp.match[0:75] + '...')\n            print(hsp.sbjct[0:75] + '...')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# `SeqRecord` and `SeqIO`\n\nOften, sequences have some additional information associated to them. \n\nA good example was the BLAST exercise which gave us sequences associated to specific organisms, and genomic locations.\n\nIdeally we would want to be able to keep this information along with our `Seq` if we have it.\n\nBioPython lets us do this with the `SeqRecord` and `SeqIO` classes.\n\n\n## Parsing with `SeqIO`\n\nThe `SeqIO` class which stands for Sequence Input/Output lets us read and write from various sequence annotation file formats which are common in biological databases.\n\nI went to the [GenBank](https://www.ncbi.nlm.nih.gov/genbank/) sequence database and looked for some sequence related to the Bubonic Plague (Yersinia Pestis bacteria) [here](https://www.ncbi.nlm.nih.gov/nuccore/NZ_ADRZ01000932.1?report=fasta).\n\n![](https://www.nationalgeographic.com/content/dam/science/photos/000/033/3338.ngsversion.1492437604403.adapt.676.1.jpg)\n(\"The Triumph of Death by  Pieter Bruegel)\n\nWe get a `fasta` file which is very common for sequence annotations so BioPython can parse it for us automatically.\n\nThe annotation for a FASTA sequence is typically held in the header:\n\n```\n>NZ_ADRZ01000932.1 Yersinia pestis biovar Antiqua str. E1979001 Contig_E1979001_19275, whole genome shotgun sequence\n```\n\nThe `SeqIO.parse` takes a path to a file and a format, in this case \"fasta\" and produces an iteratror over each entry in the fasta file.\n\nEach item produced by the iterator is a `SeqRecords` object.\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"from Bio import SeqIO\n\nrecords = SeqIO.parse(\"plague.fa\", \"fasta\")\nfor r in records:\n    print(type(r))\n    print([a for a in dir(r) if not a.startswith(\"_\")])","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### `SeqRecord` Objects\n\nIf you have a file with a single fasta record, you can use the `SeqIO.read()` function.\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"record = SeqIO.read(\"single_plague.fa\", \"fasta\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"The `SeqRecords` object has some useful attributes.\n\nThe object contains a `Seq` object witih the sequence info."},{"metadata":{"trusted":true},"cell_type":"code","source":"print(record)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"print(record.seq)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"print(record.id)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"print(record.description)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We can also create our own `SeqRecord` objects."},{"metadata":{"trusted":true},"cell_type":"code","source":"from Bio.Seq import Seq\nfrom Bio.SeqRecord import SeqRecord\nfrom Bio.Alphabet import IUPAC\nrecord = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n                       IUPAC.protein),\n                   id=\"YP_025292.1\", name=\"HokC\",\n                   description=\"toxic membrane protein, small\")\nprint(record)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Then we can write it to FASTA format."},{"metadata":{"trusted":true},"cell_type":"code","source":"record.format(\"fasta\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"And then store it in a file. \n\nSo now we know how to read sequence information, as well as produce our own sequence records and store them."},{"metadata":{"trusted":true},"cell_type":"code","source":"with open(\"myfasta.fa\", \"w\") as f:\n    f.write(record.format(\"fasta\"))\n    ","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Making Multiple Sequence Alignments and Phylogenetic Trees\n\nAnother thing biologists like to do is align (essentially, compare) sequences.\n\nIf we line up two sequences on top of each other, we can look at the parts where they are different and infer many useful pieces of information.\n\nOne of these pieces of information is the phylogeny.\n\ni.e. we can take a guess at how evolution took place to explain a particular set of observed sequences.\n\n\n\n### Step 1: Obtain some sequences and align them"},{"metadata":{"trusted":true},"cell_type":"code","source":"import random\n\n#generate some random sequences based on a random seed sequence\nmut = 0.1\nBASES = {\"A\", \"C\", \"G\", \"T\"}\nseqs = []\nseed = \"\".join([random.choice(list(BASES)) for _ in range(25 + random.choice([-1, 0, 1]))])\n\nseq_records = []\n\nfor i in range(20):\n    new_seq = \"\".join([b if random.random() > mut else random.choice(list(BASES - {b}))\\\n        for b in seed])\n    #create a seq record\n    seq_records.append(SeqRecord(Seq(new_seq), id=str(i)))\n    \nprint(seq_records)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Now that we have `SeqRecord` objects we can align them.\n\nTo do so, we need to put them in a `fasta` file.\n\n### Step 2: Align the sequences"},{"metadata":{"trusted":true},"cell_type":"code","source":"#write our sequences to fasta format\nwith open(\"seqs.fasta\", \"w\") as f:\n    SeqIO.write(seq_records, f, \"fasta\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"The next step is to actually align these sequences.\n\nBioPython has its own alignment functionalities but it requires some more installations.\n\nSo let's just use an [online tool](https://www.ebi.ac.uk/Tools/msa/muscle/) which takes a input a `fasta` file which we just created.\n\n### Step 3: Load the resulting alignment"},{"metadata":{"trusted":true},"cell_type":"code","source":"#open the alignmnent file\nfrom Bio import AlignIO\nwith open(\"aln.clustal\", \"r\") as aln:\n    #use AlignIO to read the alignment file in 'clustal' format\n    alignment = AlignIO.read(aln, \"clustal\")","execution_count":15,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We can parse the alignment with the `AlignIO` module from BioPython.\n\nI won't go into detail on this module but it has a lot of useful funcionality for dealing with sequence alignments.\n"},{"metadata":{},"cell_type":"markdown","source":"### Step 4: Use the alignment to obtain a \"distance\" between all pairs of sequences\n\nUsing the parsed alignment, we can get the distance (or difference) between all the sequences.\n\nThis just tells us, for each pair of sequences, how different they are."},{"metadata":{"trusted":true},"cell_type":"code","source":"from Bio.Phylo.TreeConstruction import DistanceCalculator\n\n#calculate the distance matrix\ncalculator = DistanceCalculator('identity')\n#adds distance matrix to the calculator object and returns it\ndm = calculator.get_distance(alignment)\nprint(dm)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Step 5: Construct a tree from the set of distances\n\nAnd finally, we can construct a phylogenetic tree from the pairwise distances between all the sequences."},{"metadata":{"trusted":true},"cell_type":"code","source":"from Bio.Phylo.TreeConstruction import DistanceTreeConstructor\n\n#initialize a DistanceTreeConstructor object based on our distance calculator object\nconstructor = DistanceTreeConstructor(calculator)\n\n#build the tree\nupgma_tree = constructor.build_tree(alignment)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"And let's use the `Phylo` module to visualize the result!"},{"metadata":{"trusted":true},"cell_type":"code","source":"from Bio import Phylo\nimport pylab\n#draw the tree\nPhylo.draw(upgma_tree)","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"},"language_info":{"name":"python","version":"3.7.6","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":2}